File: NEW CODES/LP-II/AI/Main_Code/Prac 4 (N Queens Problem)/NQueensProblem.py
class NQueensProblem:
    def __init__(self, n):
        self.queens = [0] * n
        self.numSolutions = 0

    def solve(self):
        self.solve_helper(0)

    def solve_helper(self, row):
        if row == len(self.queens):
            self.numSolutions += 1
            self.print_solution()
        else:
            for col in range(len(self.queens)):
                self.queens[row] = col
                if self.is_valid(row, col):
                    self.solve_helper(row + 1)

    def is_valid(self, row, col):
        for i in range(row):
            diff = abs(self.queens[i] - col)
            if diff == 0 or diff == row - i:
                return False
        return True

    def print_solution(self):
        if self.numSolutions == 1:
            print("Solution: ", end="")
            for i in range(len(self.queens)):
                print(self.queens[i], end=" ")
            print()
            print("The Matrix Representation:")
            arr = [[0] * len(self.queens) for _ in range(len(self.queens))]
            for i in range(len(self.queens)):
                for j in range(len(self.queens)):
                    if j == self.queens[i]:
                        arr[i][j] = 1
            for i in range(len(self.queens)):
                for j in range(len(self.queens)):
                    print(arr[i][j], end=" ")
                print()

if __name__ == "__main__":
    n = int(input("Enter N Queens Problem: "))
    NQueensProblem = NQueensProblem(n)
    NQueensProblem.solve()

"""
Enter N Queens Problem: 4  
Solution: 1 3 0 2
The Matrix Representation:  
0 1 0 0
0 0 0 1
1 0 0 0
0 0 1 0
Enter N Queens Problem: 6  
Solution: 1 3 5 0 2 4
The Matrix Representation:
0 1 0 0 0 0
0 0 0 1 0 0
0 0 0 0 0 1
1 0 0 0 0 0
0 0 1 0 0 0
0 0 0 0 1 0
Enter N Queens Problem: 8  
Solution: 0 4 7 5 2 6 1 3  
The Matrix Representation:  1 0 0 0 0 0 0 0
0 0 0 0 1 0 0 0
0 0 0 0 0 0 0 1
0 0 0 0 0 1 0 0
0 0 1 0 0 0 0 0
0 0 0 0 0 0 1 0
0 1 0 0 0 0 0 0
0 0 0 1 0 0 0 0
"""

File: NEW CODES/LP-II/AI/Main_Code/Prac 4 (N Queens Problem)/NQueensProblem.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class NQueensProblem {
private:
    vector<int> queens;
    int numSolutions;

public:
    NQueensProblem(int n) : queens(n), numSolutions(0) {}

    void solve() {
        solve(0);
    }

private:
    void solve(int row) {
        if (row == queens.size()) {
            numSolutions++;
            printSolution();
        } else {
            for (int col = 0; col < queens.size(); col++) {
                queens[row] = col;
                if (isValid(row, col)) {
                    solve(row + 1);
                }
            }
        }
    }

    bool isValid(int row, int col) {
        for (int i = 0; i < row; i++) {
            int diff =  abs(queens[i] - col);
            if (diff == 0 || diff == row - i) {
                return false;
            }
        }
        return true;
    }

    void printSolution() {
        if (numSolutions == 1) {
             cout << "Solution: ";
            for (int i = 0; i < queens.size(); i++) {
                 cout << queens[i] << " ";
            }
             cout <<  endl;
             cout << "The Matrix Representation:" <<  endl;
             vector< vector<int>> arr(queens.size(),  vector<int>(queens.size(), 0));
            for (int i = 0; i < queens.size(); i++) {
                arr[i][queens[i]] = 1;
            }
            for (int i = 0; i < queens.size(); i++) {
                for (int j = 0; j < queens.size(); j++) {
                     cout << arr[i][j] << " ";
                }
                 cout <<  endl;
            }
        }
    }
};

int main() {
    cout << "Enter N Queens Problem: ";
    int n;
    cin >> n;

    NQueensProblem nQueensProblem(n);
    nQueensProblem.solve();

    return 0;
}


/*Enter N Queens Problem: 4  
Solution: 1 3 0 2
The Matrix Representation:  
0 1 0 0
0 0 0 1
1 0 0 0
0 0 1 0
Enter N Queens Problem: 6  
Solution: 1 3 5 0 2 4
The Matrix Representation:
0 1 0 0 0 0
0 0 0 1 0 0
0 0 0 0 0 1
1 0 0 0 0 0
0 0 1 0 0 0
0 0 0 0 1 0
Enter N Queens Problem: 8  
Solution: 0 4 7 5 2 6 1 3  
The Matrix Representation:  1 0 0 0 0 0 0 0
0 0 0 0 1 0 0 0
0 0 0 0 0 0 0 1
0 0 0 0 0 1 0 0
0 0 1 0 0 0 0 0
0 0 0 0 0 0 1 0
0 1 0 0 0 0 0 0
0 0 0 1 0 0 0 0
*/
